x/ibc/07-tendermint/types/upgrade.go

```diff
@@ -0,0 +1,58 @@
+package types
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+	commitmenttypes "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment/types"
+	"github.com/cosmos/cosmos-sdk/x/ibc/exported"
+)
+
+// VerifyUpgrade checks if the upgraded client has been committed by the current client
+// It will zero out all client-specific fields (e.g. TrustingPeriod and verify all data
+// in client state that must be the same across all valid Tendermint clients for the new chain.
+// VerifyUpgrade will return an error if:
+// - the upgradedClient is not a Tendermint ClientState
+// - the height of upgraded client is not greater than that of current client
+// - the latest height of the new client does not match the height in committed client
+// - any Tendermint chain specified parameter in upgraded client such as ChainID, UnbondingPeriod,
+//   and ProofSpecs do not match parameters set by committed client
+func (cs ClientState) VerifyUpgrade(
+	ctx sdk.Context, cdc codec.BinaryMarshaler, clientStore sdk.KVStore,
+	upgradedClient exported.ClientState, proofUpgrade []byte,
+) error {
+	if cs.UpgradePath == nil {
+		return sdkerrors.Wrap(clienttypes.ErrInvalidUpgradeClient, "cannot upgrade client, no upgrade path set")
+	}
+
+	if !upgradedClient.GetLatestHeight().GT(cs.GetLatestHeight()) {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidHeight, "upgrade client height %s must be greater than current client height %s",
+			upgradedClient.GetLatestHeight(), cs.GetLatestHeight())
+	}
+
+	if len(proofUpgrade) == 0 {
+		return sdkerrors.Wrap(commitmenttypes.ErrInvalidProof, "proof of upgrade is empty")
+	}
+
+	var merkleProof commitmenttypes.MerkleProof
+	if err := cdc.UnmarshalBinaryBare(proofUpgrade, &merkleProof); err != nil {
+		return sdkerrors.Wrapf(commitmenttypes.ErrInvalidProof, "could not unmarshal merkle proof: %v", err)
+	}
+
+	// counterparty chain must commit the upgraded client with all client-customizable fields zeroed out
+	// at the upgrade path specified by current client
+	committedClient := upgradedClient.ZeroCustomFields()
```

Comment:
Where do we check that the non-zeroed-out fields correspond to the old (current) fields? The relayer shouldn't be choosing these

Reply:
It's not generally possible because the client state structure may have changed. I just check that the chosen fields are consistent with unchosen fields.

Reply:
Can't the relayer change the trusting ratio or something then? Are we preventing that somewhere?

Reply:
I think the issue here is the client state could change to something radically different (a solo machine client is used in testing) which may not have a trusting ratio

Reply:
hmm, that would be alright, but we still can't let the relayer change these fields in e.g. a Tendermint to Tendermint upgrade